### PR TITLE
[ci] Travis CI doesn't need to build apidocs any more

### PR DESCRIPTION
**Pull request by eduardoj**
Tests for apidocs controller has been changed to work without a previous build
of apidocs.

### CODE REVIEW

**Comment by evanrolfe** [comment]
Whats the reason for this change?

**Comment by eduardoj** [comment]
The reason for this change is to have only controller tests in the apidocs controller test file.

With this change it will not be needed to build apidocs. This controller test file will not test if apidocs is built without problems, or if an apidocs build process will create files or not. It will test only all the controller actions, despite apidocs are built or not. It is not needed. Travis CI test will be faster.

File: src/api/spec/controllers/webui/apidocs_controller_spec.rb [hunk 1]
```diff
@@ -2,26 +2,40 @@
...
-      let!(:old_location) { CONFIG['apidocs_location'] }
-
       before do
-        CONFIG['apidocs_location'] = '/your/mom'
-      end
-
-      after do
-        CONFIG['apidocs_location'] = old_location
+        CONFIG['apidocs_location'] = 'non/existent/subdirectory'
+        allow(File).to receive(:exist?).and_return(false)
```
Comment (evanrolfe) [inline src/api/spec/controllers/webui/apidocs_controller_spec.rb]: Is it necessary to mock File.exists? I'd prefer if it wasn't mocked so as not to make test too specific to the implementation of the controller.

File: dist/ci/travis_before_script.sh [hunk 2]
```diff
@@ -31,8 +31,3 @@ popd
 
 # travis rvm can not deal with our extended executable names
 sed -i 1,1s,\.ruby2\.4,, src/api/{script,bin}/* docs/api/restility/bin/*
-
-echo "Build apidocs"
-pushd docs/api
```
Comment (evanrolfe) [inline dist/ci/travis_before_script.sh]: I'm not too familiar with how the api docs are built so I may be wrong, but don't we want to include this in the script? So travis checks that the api docs build ok? Otherwise if they are broken we might not find out until somebody complains about them..

Comment (eduardoj) [inline dist/ci/travis_before_script.sh]: If Travis must check that apidocs builds without problems it will not be in a controller test. Or will it be?

**Comment by eduardoj** [comment]
> So basically the controller spec tests less things now? I don't see how thats better?

The controller spec tests now test the controller.
 
> This change reduces the coverage of the test and also adds more implementation-specific setup code to the test which makes it harder to change the internals of the controller without changing the test.
> 
> I don't like it.. Sorry!

The coverage remains the same, at 65.151%  . Maybe you need to reload the page.

Yes, the test is more implementation specific, but, as a controller test, relays only on the controller, and not on the apidocs build process.

File: src/api/spec/controllers/webui/apidocs_controller_spec.rb (hunk 1 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/apidocs_controller_spec.rb]: In this case it is necessary to be sure 'non/existent/subdirectory' does not exist.

What could be your alternative, without being mocked?

**Comment by evanrolfe** [comment]
> The controller spec tests now test the controller.
> ...the test is more implementation specific, but, as a controller test, relays only on the controller, and not on the apidocs build process.

Yeah this is not a good thing. [Here](http://david.heinemeierhansson.com/2014/test-induced-design-damage.html) is a good explanation of why thats the case.

> The coverage remains the same, at 65.151% . Maybe you need to reload the page.

When I say the coverage has decreased I mean that before if somebody made a change that broke the api docs build, we would see that as an error in travis (from the controller spec). Now we wont. I'm pretty sure that coveralls doesn't include the api docs building in its calculation so the number is the same.

**Comment by hennevogel** [comment]
Sorry @eduardoj I also don't like it, for the same reasons.

**Comment by mdeniz** [comment]
I think that there are two separate things here, one is to test the controller properly (execution flow and outcome) and the other is to test if the apidocs will be properly setup. If we keep it as before the controller test will have a dependency on a file (not if apidocs is installed o executed properly, just a file somewhere) that supposedly is set by a third part tool (I prefer to have a test dependent on a file I manage than on a tool that not).

So, rather than mixing both in a controller test I would prefer to test the apidocs setup in a different place. And be focus in the controller test what the action actually does, (so, checking that a file exist to show it or redirect back with an error flash), and checking the outcome (so, the file I set before in the filesystem is what I got when run a get to those routes).

After that... @evanrolfe and @hennevogel how do you think it should be tested if we want to do it without having a dependency on a working setup of apidocs?

BTW, using a global directory location to test the broken setup is not the best, I mean doing this
```ruby
CONFIG['apidocs_location'] = '/your/mom'
```
It at least should be changed to something inside our code tmp dir.

**Comment by evanrolfe** [comment]
@mdeniz it seems like this is coming down to two different testing approaches. Eduardo and your way is _"test the controller as I've designed it to work"_ meaning you only test the code thats written inside the controller and nothing else, you assume that the api docs is working exactly as the controller code expects it to.

My way and Henne's is _"test the controller as it will work in reality"_. We do not assume that the api docs works correctly because if the api docs do not work then the controller will fail too, and if the controller will fail then so should the controller test. It doesn't matter what caused the controller to fail, all that matters is that it fails.

I highly recommend watching this video, its about unit testing models but the same principles apply here and she talks about very similar situations to this api docs controller test:
https://www.youtube.com/watch?v=URSWYvyc42M

If you watch that talk you will see that the ApidocsController#index and #file methods fall into the "Incoming Query Message" category, therefore the best approach (according to her) is to:
- "assert on the result"
- "sight along the edges of the space capsules"
- "test the interface and not the implementation"

Not that I think she is 100% right about everything but I find her way of testing pretty convincing as the way to go..

**Comment by mdeniz** [comment]
@evanrolfe already saw that talk time ago, is awesome. Thanks for refreshing me that I re-watched that 👍 

But my point is, the controller is just sending a file in a "random" location (the configured one in production for sure), this doesn't ensure you that apidocs worked at all, it only ensures you that there are files in the system that match the location (we don't perform a library call or something that may fail).

As the "Incoming Query Message" category says "assert on the result" that in this case means test when you send a filename that is there or not that can look like is implementation dependent... but not, is what I have in return (a file or an error/redirect). That action relies on the message params (none for #index or filename for #file action) and in the state of the system...

Maybe @eduardoj implementation of the test using a stub for File.exists? is too much dependent on the insides of the action but I'm not saying that I agree on having that, maybe is better just having a file there and let the controller do the check. But at least you are sure that the file is there, you don't rely on the fact that apidoc properly worked once. You are sure how the action will behave because you controll the input (params + system state) and not the apidocs execution.

For instance, if you have your dev machine working and you removed the index.html of the apidocs by accident then you will have a controller spec failing when the action code haven't changed. I rather prefer to see it failing on Travis because the apidocs didn't work properly.

To summarize it, I don't want to remove the setup of the apidocs in Travis to see it failing if is not working. And I prefer to setup the files that will be my state for testing the actions. In the broken_setup case we should also not use just a path that can exist, we should be sure that not by maybe removing it before running the test.

Ufff... tooo long coment hahaha nice weekend BTW

**Comment by evanrolfe** [comment]
>Maybe @eduardoj implementation of the test using a stub for File.exists? is too much dependent on the insides of the action but I'm not saying that I agree on having that, maybe is better just having a file there and let the controller do the check. But at least you are sure that the file is there, you don't rely on the fact that apidoc properly worked once.

Yeah I see your point about this, since the api docs setup is only ran once at setup.. With that in mind I would also be okay with this PR if `File.exists?` wasn't stubbed. I don't think its necessary is it?

**Comment by evanrolfe** [comment]
LGTM (when travis passes)
